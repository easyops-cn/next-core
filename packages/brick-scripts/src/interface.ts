export type FileWithContent = [string, string];

export interface AskFlags {
  type?: TargetType;
  provider?: "string";
}

export enum TargetType {
  A_NEW_BRICK = "brick",
  A_NEW_EDITOR_BRICK = "editor-brick",
  A_NEW_PACKAGE_OF_BRICKS = "bricks",
  A_NEW_PACKAGE_OF_LIBS = "libs",
  // A_NEW_PACKAGE_OF_MICRO_APPS = "micro-apps",
  A_NEW_CUSTOM_PROVIDER = "custom-provider",
  A_NEW_CUSTOM_PROCESSOR = "custom-processor",
  A_NEW_PACKAGE_OF_PROVIDERS = "providers",
  A_NEW_PACKAGE_OF_DLL = "dll",
  // TRANSFORM_A_MICRO_APP = "transform",
  A_NEW_CUSTOM_TEMPLATE = "custom-template",
  // A_NEW_LEGACY_TEMPLATE = "legacy-template",
  // A_NEW_PACKAGE_OF_LEGACY_TEMPLATES = "legacy-templates",
  // I18N_PATCH_A_PACKAGE_OF_LEGACY_TEMPLATES = "i18n-patch",
}

export const TargetTypeDisplay = {
  [TargetType.A_NEW_BRICK]: "a new brick",
  [TargetType.A_NEW_EDITOR_BRICK]: "a new editor brick",
  [TargetType.A_NEW_PACKAGE_OF_BRICKS]: "a new package of bricks",
  [TargetType.A_NEW_PACKAGE_OF_LIBS]: "a new package of libs",
  // [TargetType.A_NEW_PACKAGE_OF_MICRO_APPS]: "a new package of micro-apps",
  [TargetType.A_NEW_CUSTOM_PROVIDER]: "a new custom provider",
  [TargetType.A_NEW_CUSTOM_PROCESSOR]: "a new custom processor",
  [TargetType.A_NEW_PACKAGE_OF_PROVIDERS]: "a new package of providers",
  [TargetType.A_NEW_PACKAGE_OF_DLL]: "a new package of dll",
  // [TargetType.TRANSFORM_A_MICRO_APP]: "transform a micro-app",
  [TargetType.A_NEW_CUSTOM_TEMPLATE]: "a new custom template",
  // [TargetType.A_NEW_LEGACY_TEMPLATE]: "a new legacy template",
  // [TargetType.A_NEW_PACKAGE_OF_LEGACY_TEMPLATES]:
  //   "a new package of legacy templates",
  // [TargetType.I18N_PATCH_A_PACKAGE_OF_LEGACY_TEMPLATES]:
  //   "i18n-patch a package of legacy templates",
};
