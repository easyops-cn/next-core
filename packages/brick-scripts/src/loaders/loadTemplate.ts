import fs from "fs-extra";
import path from "path";
import { exec } from "child_process";
import klawSync from "klaw-sync";
import * as changeCase from "change-case";
import rp from "request-promise-native";
import { getEasyopsConfig } from "@next-core/repo-config";
import { FileWithContent, TargetType } from "../interface";

// `tsc` will compile files which `import` or `require`,
// thus, we read file content instead of importing.
const packageJson = JSON.parse(
  fs.readFileSync(path.join(__dirname, "../../package.json"), "utf8")
);
const { templateDependencies } = packageJson;
const brickContainerVersion =
  templateDependencies["@next-core/brick-container"];

const workspacePackageJson = JSON.parse(
  fs.readFileSync(path.join(process.cwd(), "package.json"), "utf8")
);
const workspaceHomepage = workspacePackageJson.homepage;
const workspaceLicense = workspacePackageJson.license || "UNLICENSED";
let userName = "bot";
exec("git config user.name", { encoding: "utf8" }, function (err, value) {
  if (err) return;
  userName = value.trim();
});

function escapeRegExp(str: string): string {
  return str.replace(/([.*+?^=!:${}()|[\]/\\])/g, "\\$1");
}

function replaceAll(str: string, find: string, replace: string): string {
  return str.replace(new RegExp(escapeRegExp(find), "g"), replace);
}

function replaceFileContent(
  filePath: string,
  translations: Record<string, string>
): string {
  const content = fs.readFileSync(filePath, "utf8");
  const contentFinds = Object.keys(translations).filter((key) =>
    content.includes(key)
  );
  if (contentFinds.length > 0) {
    return contentFinds.reduce(
      (acc, find) => replaceAll(acc, find, translations[find]),
      content
    );
  }
  return content;
}

export async function loadTemplate({
  targetType,
  packageName,
  brickName,
  processorName,
  targetRoot,
}: {
  targetType: TargetType;
  packageName: string;
  brickName: string;
  processorName: string;
  targetRoot: string;
}): Promise<FileWithContent[]> {
  const { useLocalSdk, usePublicScope } = getEasyopsConfig();

  const npmScopeOfBricks = usePublicScope ? "@next-bricks" : "@bricks";
  const npmScopeOfMicroApps = usePublicScope
    ? "@next-micro-apps"
    : "@micro-apps";
  const npmScopeOfTemplates = usePublicScope
    ? "@next-legacy-templates"
    : "@templates";
  const npmScopeOfLibs = usePublicScope ? "@next-libs" : "@libs";
  const npmScopeOfSdk = usePublicScope ? "@next-sdk" : "@sdk";
  const repoOrgUrl = usePublicScope
    ? "https://github.com/easyops-cn"
    : "https://git.easyops.local/anyclouds";
  const repoGitUrl = usePublicScope
    ? "github.com:easyops-cn"
    : "git.easyops.local:anyclouds";
  const npmRegistryUrl = usePublicScope
    ? "https://registry.npmjs.org"
    : "https://registry.npm.easyops.local";

  const targetMap: { [key: string]: string } = {
    [TargetType.A_NEW_BRICK]: "brick",
    [TargetType.A_NEW_EDITOR_BRICK]: "editor-brick",
    [TargetType.A_NEW_CUSTOM_TEMPLATE]: "custom-template",
    [TargetType.A_NEW_PACKAGE_OF_BRICKS]: "bricks-pkg",
    [TargetType.A_NEW_PACKAGE_OF_LIBS]: "libs-pkg",
    [TargetType.A_NEW_CUSTOM_PROVIDER]: "custom-provider",
    [TargetType.A_NEW_CUSTOM_PROCESSOR]: "custom-processor",
    [TargetType.A_NEW_PACKAGE_OF_PROVIDERS]: "providers-pkg",
    [TargetType.A_NEW_PACKAGE_OF_DLL]: "dll-pkg",
  };
  const templateDirOfFileName = targetMap[targetType];
  const templateRoot = path.join(__dirname, "../../template");
  let templateDir = path.join(templateRoot, templateDirOfFileName);

  const ignores = [".DS_Store"];
  let sdkName: string;
  if (targetType === TargetType.A_NEW_PACKAGE_OF_PROVIDERS) {
    // Providers 库直接使用构件库的模板文件，但忽略其中部分文件。
    templateDir = path.join(
      templateRoot,
      targetMap[TargetType.A_NEW_PACKAGE_OF_BRICKS]
    );
    // Providers 库不需要 `src` 及 `webpack.editors.js`。
    ignores.push("template/bricks-pkg/src/");
    ignores.push("template/bricks-pkg/webpack.editors.js");
    sdkName = `${packageName.replace(/^providers-of-/, "")}-sdk`;
  }

  const translations: Record<string, string> = {
    "$npm-scope-of-bricks$": npmScopeOfBricks,
    "$npm-scope-of-micro-apps$": npmScopeOfMicroApps,
    "$npm-scope-of-templates$": npmScopeOfTemplates,
    "$npm-scope-of-libs$": npmScopeOfLibs,
    "$npm-scope-of-sdk$": npmScopeOfSdk,
    "$repo-org-url$": repoOrgUrl,
    "$repo-git-url$": repoGitUrl,
    "$workspace-homepage$": workspaceHomepage,
    "$kebab-package-name$": packageName,
    $PascalPackageName$: changeCase.pascalCase(packageName),
    "$Title Package Name$": changeCase.capitalCase(packageName),
    $CONSTANT_PACKAGE_NAME$: changeCase.constantCase(packageName),
    $camelCasePackageName$: changeCase.camelCase(packageName),
    $PascalBrickName$: changeCase.pascalCase(brickName),
    "$kebab-brick-name$": `${packageName}.${brickName}`,
    "$kebab-brick-last-name$": brickName,
    "$kebab-custom-provider-name$": `${packageName}.provider-${brickName}`,
    "$generator.version$": `v${packageJson.version}`,
    "$brick.container.version$": brickContainerVersion,
    "$kebab-sdk-name$": sdkName,
    "$kebab-username$": userName,
    $camelProcessorName$: changeCase.camelCase(processorName),
    "$open-source-license$": workspaceLicense,
    "$brick-package-type$":
      targetType === TargetType.A_NEW_PACKAGE_OF_PROVIDERS
        ? "providers"
        : "bricks",
  };

  const filter = (src: string): boolean =>
    ignores.every((item) => !src.includes(item));

  const templateGroups: any = [
    {
      templateDir,
      targetDir: targetRoot,
      files: klawSync(templateDir, {
        depthLimit: 4,
        nodir: true,
        filter: (item) => filter(item.path),
      }),
    },
  ];

  if (targetType === TargetType.A_NEW_PACKAGE_OF_BRICKS && brickName) {
    // Also create a new brick for the new bricks-package
    const brickTemplateDir = path.join(templateRoot, "brick");
    templateGroups.push({
      templateDir: brickTemplateDir,
      targetDir: path.join(targetRoot, "src", brickName),
      files: klawSync(brickTemplateDir, {
        depthLimit: 2,
        nodir: true,
        filter: (item) => filter(item.path),
      }),
    });
  } else if (targetType === TargetType.A_NEW_PACKAGE_OF_PROVIDERS) {
    // Providers 库还有专属的一些模板文件。
    const brickTemplateDir = path.join(templateRoot, "providers-pkg");
    templateGroups.push({
      templateDir: brickTemplateDir,
      targetDir: targetRoot,
      files: klawSync(brickTemplateDir, {
        depthLimit: 2,
        nodir: true,
        filter: (item) => filter(item.path),
      }),
    });

    let sdkVersion = "1.0.0";
    const getVersion = async () => {
      try {
        const sdkPackage = await rp({
          url: `${npmRegistryUrl}/${npmScopeOfSdk}/${sdkName}`,
          json: true,
          strictSSL: false,
        });
        sdkVersion = `^${sdkPackage["dist-tags"].latest}`;
      } catch {
        sdkVersion = "FETCH LATEST VERSION ERROR";
      }
    };

    if (!useLocalSdk) {
      await getVersion();
    } else {
      const sdkPackagePath = path.resolve(
        targetRoot,
        "..",
        "..",
        "sdk",
        sdkName,
        "package.json"
      );
      if (fs.existsSync(sdkPackagePath)) {
        sdkVersion = JSON.parse(
          fs.readFileSync(sdkPackagePath, "utf8")
        ).version;
      }
    }

    translations["$sdk.version$"] = sdkVersion;
  }

  const files: FileWithContent[] = templateGroups.reduce(
    (acc: FileWithContent[], group: any) =>
      acc.concat(
        group.files.map((file: any) => {
          const content = replaceFileContent(file.path, translations);

          let realPath = file.path;
          const basename = path.basename(file.path);
          const filenameFinds = Object.keys(translations).filter((key) =>
            basename.includes(key)
          );
          if (filenameFinds.length > 0) {
            realPath = path.join(
              path.dirname(realPath),
              filenameFinds.reduce(
                (acc, find) => replaceAll(acc, find, translations[find]),
                basename
              )
            );
          }

          return [
            path.join(
              group.targetDir,
              path.relative(group.templateDir, realPath)
            ),
            content,
          ];
        })
      ),
    []
  );

  if (
    [
      TargetType.A_NEW_PACKAGE_OF_BRICKS,
      TargetType.A_NEW_PACKAGE_OF_LIBS,
      TargetType.A_NEW_PACKAGE_OF_PROVIDERS,
      TargetType.A_NEW_PACKAGE_OF_DLL,
    ].includes(targetType)
  ) {
    files.push([
      path.join(targetRoot, "package.json"),
      replaceFileContent(
        path.join(templateRoot, `${targetMap[targetType]}.json`),
        translations
      ),
    ]);
  }

  return files;
}
