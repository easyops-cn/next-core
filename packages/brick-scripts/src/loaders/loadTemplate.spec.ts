import path from "path";
import { getEasyopsConfig } from "@next-core/repo-config";
import { loadTemplate } from "./loadTemplate";
import { FileWithContent, TargetType } from "../interface";

jest.mock("@next-core/repo-config");
(getEasyopsConfig as jest.Mock).mockReturnValue({
  useLocalSdk: true,
});

jest.mock("request-promise-native");

function ignoreVersionRelatedFiles(
  files: FileWithContent[]
): FileWithContent[] {
  return files.filter(([filePath]) => {
    const fileName = path.basename(filePath);
    return !["package.json", "README.md", "package.conf.yaml"].includes(
      fileName
    );
  });
}

describe("loadTemplate", () => {
  it("should create a new brick", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_BRICK,
      packageName: "for-good",
      brickName: "for-better",
      processorName: "",
      targetRoot: "dist",
    });
    expect(files).toMatchSnapshot();
  });
  it("should create a new editor brick", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_EDITOR_BRICK,
      packageName: "for-good",
      brickName: "for-better",
      processorName: "",
      targetRoot: "dist",
    });
    expect(files).toMatchSnapshot();
  });
  it("should create a new custom template", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_CUSTOM_TEMPLATE,
      packageName: "for-good",
      brickName: "for-better",
      processorName: "",
      targetRoot: "dist",
    });
    expect(files).toMatchSnapshot();
  });
  it("should create a new package of bricks", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_PACKAGE_OF_BRICKS,
      packageName: "for-good",
      brickName: "for-better",
      processorName: "",
      targetRoot: "dist",
    });
    const otherFiles = ignoreVersionRelatedFiles(files);
    expect(otherFiles).toMatchSnapshot();
  });
  it("should create a new package of bricks without any bricks", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_PACKAGE_OF_BRICKS,
      packageName: "for-good",
      brickName: "",
      processorName: "",
      targetRoot: "dist",
    });
    const otherFiles = ignoreVersionRelatedFiles(files);
    expect(otherFiles).toMatchSnapshot();
  });
  it("should create a new package of libs", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_PACKAGE_OF_LIBS,
      packageName: "for-good",
      brickName: "",
      processorName: "",
      targetRoot: "dist",
    });
    const otherFiles = ignoreVersionRelatedFiles(files);
    expect(otherFiles).toMatchSnapshot();
  });
  it("should create a new custom provider", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_CUSTOM_PROVIDER,
      packageName: "for-good",
      brickName: "get-data",
      processorName: "",
      targetRoot: "dist",
    });
    expect(files).toMatchSnapshot();
  });
  it("should create a new custom processor", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_CUSTOM_PROCESSOR,
      packageName: "for-good",
      brickName: "",
      processorName: "doGood",
      targetRoot: "dist",
    });
    expect(files).toMatchSnapshot();
  });
  it("should create a new package of providers", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_PACKAGE_OF_PROVIDERS,
      packageName: "for-good",
      brickName: "",
      processorName: "",
      targetRoot: "dist",
    });
    const otherFiles = ignoreVersionRelatedFiles(files);
    expect(otherFiles).toMatchSnapshot();
  });
  it("should create a new package of dll", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_PACKAGE_OF_DLL,
      packageName: "for-good",
      brickName: "",
      processorName: "",
      targetRoot: "dist",
    });
    const otherFiles = ignoreVersionRelatedFiles(files);
    expect(otherFiles).toMatchSnapshot();
  });
});
