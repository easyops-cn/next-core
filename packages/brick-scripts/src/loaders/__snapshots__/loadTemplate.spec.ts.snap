// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadTemplate should create a new brick 1`] = `
Array [
  Array [
    "dist/ForBetter.spec.tsx",
    "import React from \\"react\\";
import { fireEvent, render, screen } from \\"@testing-library/react\\";
import \\"@testing-library/jest-dom\\";
import { ForBetter } from \\"./ForBetter\\";

describe(\\"ForBetter\\", () => {
  it(\\"should work\\", () => {
    render(<ForBetter />);
    fireEvent.click(screen.getByTestId(\\"my-brick\\"));
    expect(screen.getByTestId(\\"my-brick\\")).toHaveTextContent(
      \\"FOR_GOOD works!\\"
    );
  });
});
",
  ],
  Array [
    "dist/ForBetter.tsx",
    "import React from \\"react\\";
import { useTranslation } from \\"react-i18next\\";
import { NS_FOR_GOOD, K } from \\"../i18n/constants\\";

export function ForBetter(): React.ReactElement {
  const { t } = useTranslation(NS_FOR_GOOD);

  return (
    <div data-testid=\\"my-brick\\">{t(K.FOR_GOOD)} works!</div>
  );
}
",
  ],
  Array [
    "dist/index.spec.ts",
    "import ReactDOM from \\"react-dom\\";
import \\"./\\";

const spyOnRender = jest
  .spyOn(ReactDOM, \\"render\\")
  .mockImplementation(() => null);
const unmountComponentAtNode = jest
  .spyOn(ReactDOM, \\"unmountComponentAtNode\\")
  .mockImplementation(() => null);

describe(\\"for-good.for-better\\", () => {
  it(\\"should create a custom element\\", () => {
    const element = document.createElement(\\"for-good.for-better\\");
    expect(spyOnRender).not.toBeCalled();
    document.body.appendChild(element);
    expect(spyOnRender).toBeCalled();
    document.body.removeChild(element);
    expect(unmountComponentAtNode).toBeCalled();
  });
});
",
  ],
  Array [
    "dist/index.tsx",
    "import React from \\"react\\";
import ReactDOM from \\"react-dom\\";
import { BrickWrapper, UpdatingElement } from \\"@next-core/brick-kit\\";
import { ForBetter } from \\"./ForBetter\\";

/**
 * @id for-good.for-better
 * @author bot
 * @history
 * 1.x.0: 新增构件 \`for-good.for-better\`
 * @docKind brick
 * @noInheritDoc
 */
export class ForBetterElement extends UpdatingElement {
  connectedCallback(): void {
    // Don't override user's style settings.
    // istanbul ignore else
    if (!this.style.display) {
      this.style.display = \\"block\\";
    }
    this._render();
  }

  disconnectedCallback(): void {
    ReactDOM.unmountComponentAtNode(this);
  }

  protected _render(): void {
    // istanbul ignore else
    if (this.isConnected) {
      ReactDOM.render(
        <BrickWrapper>
          <ForBetter />
        </BrickWrapper>,
        this
      );
    }
  }
}

customElements.define(\\"for-good.for-better\\", ForBetterElement);
",
  ],
]
`;

exports[`loadTemplate should create a new custom processor 1`] = `
Array [
  Array [
    "dist/doGood.spec.ts",
    "import { doGood } from \\"./doGood\\";

jest.mock(\\"@next-core/brick-kit\\", () => ({
  getRuntime: () => ({
    registerCustomProcessor: jest.fn(),
  }),
}));

describe(\\"doGood\\", () => {
  it(\\"should work\\", () => {
    expect(doGood()).toEqual(undefined);
  });
});
",
  ],
  Array [
    "dist/doGood.ts",
    "import { getRuntime } from \\"@next-core/brick-kit\\";

export function doGood(): any {
  // ...
}

getRuntime().registerCustomProcessor(
  \\"forGood.doGood\\",
  doGood
);
",
  ],
]
`;

exports[`loadTemplate should create a new custom provider 1`] = `
Array [
  Array [
    "dist/GetData.spec.ts",
    "import {
  GetData,
  GetDataParams,
} from \\"./GetData\\";

describe(\\"GetData\\", () => {
  it.each<[GetDataParams, any]>([[{}, null]])(
    \\"GetData(%j) should work\\",
    async (params, result) => {
      expect(await GetData(params)).toEqual(result);
    }
  );
});
",
  ],
  Array [
    "dist/GetData.ts",
    "import { createProviderClass } from \\"@next-core/brick-utils\\";

export interface GetDataParams {}

export function GetData(
  params: GetDataParams
): Promise<unknown> {
  return null;
}

customElements.define(
  \\"for-good.provider-get-data\\",
  createProviderClass(GetData)
);
",
  ],
]
`;

exports[`loadTemplate should create a new custom template 1`] = `
Array [
  Array [
    "dist/for-better.ts",
    "import { getRuntime } from \\"@next-core/brick-kit\\";

getRuntime().registerCustomTemplate(\\"for-good.for-better\\", {
  proxy: {},
  bricks: [],
});
",
  ],
]
`;

exports[`loadTemplate should create a new package of bricks 1`] = `
Array [
  Array [
    "dist/jest.config.js",
    "const { jestConfigFactory } = require(\\"@next-core/jest-config-factory\\");

module.exports = jestConfigFactory({
  standalone: true,
  cwd: __dirname,
});
",
  ],
  Array [
    "dist/rollup.snippets.js",
    "import { rollupFactoryForSnippets } from \\"@next-core/rollup-config-factory\\";

export default rollupFactoryForSnippets();
",
  ],
  Array [
    "dist/scripts/generate-version-file.js",
    "const { createVersionFile } = require(\\"@next-core/build-config-factory\\");

createVersionFile();
",
  ],
  Array [
    "dist/scripts/post-build.js",
    "const { postBuild } = require(\\"@next-core/build-config-factory\\");

postBuild(\\"bricks\\");
",
  ],
  Array [
    "dist/scripts/pre-build.js",
    "const { preBuild } = require(\\"@next-core/build-config-factory\\");

preBuild(\\"bricks\\");
",
  ],
  Array [
    "dist/scripts/rename-tar-package.js",
    "const { renameTarPackage } = require(\\"@next-core/build-config-factory\\");

renameTarPackage();
",
  ],
  Array [
    "dist/src/i18n/constants.ts",
    "export const NS_FOR_GOOD = \\"for-good\\";

export enum K {
  FOR_GOOD = \\"FOR_GOOD\\"
}

export type Locale = { [key in K]: string };
",
  ],
  Array [
    "dist/src/i18n/index.ts",
    "import i18next from \\"i18next\\";
import { NS_FOR_GOOD } from \\"./constants\\";
import en from \\"./locales/en\\";
import zh from \\"./locales/zh\\";

i18next.addResourceBundle(\\"en\\", NS_FOR_GOOD, en);
i18next.addResourceBundle(\\"zh\\", NS_FOR_GOOD, zh);
",
  ],
  Array [
    "dist/src/i18n/locales/en.ts",
    "import { Locale, K } from \\"../constants\\";

const locale: Locale = {
  [K.FOR_GOOD]: \\"For Good\\"
};

export default locale;
",
  ],
  Array [
    "dist/src/i18n/locales/zh.ts",
    "import { Locale, K } from \\"../constants\\";

const locale: Locale = {
  [K.FOR_GOOD]: \\"For Good\\"
};

export default locale;
",
  ],
  Array [
    "dist/src/index.spec.ts",
    "import i18next from \\"i18next\\";
import * as kit from \\"@next-core/brick-kit\\";

const spyOnAddResourceBundle = (i18next.addResourceBundle = jest.fn());

jest.spyOn(window.customElements, \\"define\\");

jest.spyOn(kit, \\"getRuntime\\").mockReturnValue({
  registerCustomTemplate: jest.fn(),
  registerCustomProcessor: jest.fn(),
} as any);

jest.mock(\\"./lazy-bricks\\", () => void 0);

// Use \`require\` instead of \`import\` to avoid hoisting.
require(\\"./index\\");

describe(\\"index\\", () => {
  it(\\"should add i18n resource bundle\\", () => {
    expect(spyOnAddResourceBundle).toBeCalled();
  });
});
",
  ],
  Array [
    "dist/src/index.ts",
    "import \\"./i18n\\";
",
  ],
  Array [
    "dist/tsconfig.json",
    "{
  \\"extends\\": \\"../../tsconfig.json\\",
  \\"include\\": [
    \\"./src\\",
    \\"../../declarations\\"
  ]
}
",
  ],
  Array [
    "dist/webpack.config.js",
    "const { bricks, merge } = require(\\"@next-core/webpack-config-factory\\");

const { webpackCommonFactory, webpackDevFactory, webpackProdFactory } = bricks;

module.exports = merge(
  webpackCommonFactory(),
  process.env.NODE_ENV === \\"development\\"
    ? webpackDevFactory()
    : webpackProdFactory()
);
",
  ],
  Array [
    "dist/src/for-better/ForBetter.spec.tsx",
    "import React from \\"react\\";
import { fireEvent, render, screen } from \\"@testing-library/react\\";
import \\"@testing-library/jest-dom\\";
import { ForBetter } from \\"./ForBetter\\";

describe(\\"ForBetter\\", () => {
  it(\\"should work\\", () => {
    render(<ForBetter />);
    fireEvent.click(screen.getByTestId(\\"my-brick\\"));
    expect(screen.getByTestId(\\"my-brick\\")).toHaveTextContent(
      \\"FOR_GOOD works!\\"
    );
  });
});
",
  ],
  Array [
    "dist/src/for-better/ForBetter.tsx",
    "import React from \\"react\\";
import { useTranslation } from \\"react-i18next\\";
import { NS_FOR_GOOD, K } from \\"../i18n/constants\\";

export function ForBetter(): React.ReactElement {
  const { t } = useTranslation(NS_FOR_GOOD);

  return (
    <div data-testid=\\"my-brick\\">{t(K.FOR_GOOD)} works!</div>
  );
}
",
  ],
  Array [
    "dist/src/for-better/index.spec.ts",
    "import ReactDOM from \\"react-dom\\";
import \\"./\\";

const spyOnRender = jest
  .spyOn(ReactDOM, \\"render\\")
  .mockImplementation(() => null);
const unmountComponentAtNode = jest
  .spyOn(ReactDOM, \\"unmountComponentAtNode\\")
  .mockImplementation(() => null);

describe(\\"for-good.for-better\\", () => {
  it(\\"should create a custom element\\", () => {
    const element = document.createElement(\\"for-good.for-better\\");
    expect(spyOnRender).not.toBeCalled();
    document.body.appendChild(element);
    expect(spyOnRender).toBeCalled();
    document.body.removeChild(element);
    expect(unmountComponentAtNode).toBeCalled();
  });
});
",
  ],
  Array [
    "dist/src/for-better/index.tsx",
    "import React from \\"react\\";
import ReactDOM from \\"react-dom\\";
import { BrickWrapper, UpdatingElement } from \\"@next-core/brick-kit\\";
import { ForBetter } from \\"./ForBetter\\";

/**
 * @id for-good.for-better
 * @author bot
 * @history
 * 1.x.0: 新增构件 \`for-good.for-better\`
 * @docKind brick
 * @noInheritDoc
 */
export class ForBetterElement extends UpdatingElement {
  connectedCallback(): void {
    // Don't override user's style settings.
    // istanbul ignore else
    if (!this.style.display) {
      this.style.display = \\"block\\";
    }
    this._render();
  }

  disconnectedCallback(): void {
    ReactDOM.unmountComponentAtNode(this);
  }

  protected _render(): void {
    // istanbul ignore else
    if (this.isConnected) {
      ReactDOM.render(
        <BrickWrapper>
          <ForBetter />
        </BrickWrapper>,
        this
      );
    }
  }
}

customElements.define(\\"for-good.for-better\\", ForBetterElement);
",
  ],
]
`;

exports[`loadTemplate should create a new package of bricks without any bricks 1`] = `
Array [
  Array [
    "dist/jest.config.js",
    "const { jestConfigFactory } = require(\\"@next-core/jest-config-factory\\");

module.exports = jestConfigFactory({
  standalone: true,
  cwd: __dirname,
});
",
  ],
  Array [
    "dist/rollup.snippets.js",
    "import { rollupFactoryForSnippets } from \\"@next-core/rollup-config-factory\\";

export default rollupFactoryForSnippets();
",
  ],
  Array [
    "dist/scripts/generate-version-file.js",
    "const { createVersionFile } = require(\\"@next-core/build-config-factory\\");

createVersionFile();
",
  ],
  Array [
    "dist/scripts/post-build.js",
    "const { postBuild } = require(\\"@next-core/build-config-factory\\");

postBuild(\\"bricks\\");
",
  ],
  Array [
    "dist/scripts/pre-build.js",
    "const { preBuild } = require(\\"@next-core/build-config-factory\\");

preBuild(\\"bricks\\");
",
  ],
  Array [
    "dist/scripts/rename-tar-package.js",
    "const { renameTarPackage } = require(\\"@next-core/build-config-factory\\");

renameTarPackage();
",
  ],
  Array [
    "dist/src/i18n/constants.ts",
    "export const NS_FOR_GOOD = \\"for-good\\";

export enum K {
  FOR_GOOD = \\"FOR_GOOD\\"
}

export type Locale = { [key in K]: string };
",
  ],
  Array [
    "dist/src/i18n/index.ts",
    "import i18next from \\"i18next\\";
import { NS_FOR_GOOD } from \\"./constants\\";
import en from \\"./locales/en\\";
import zh from \\"./locales/zh\\";

i18next.addResourceBundle(\\"en\\", NS_FOR_GOOD, en);
i18next.addResourceBundle(\\"zh\\", NS_FOR_GOOD, zh);
",
  ],
  Array [
    "dist/src/i18n/locales/en.ts",
    "import { Locale, K } from \\"../constants\\";

const locale: Locale = {
  [K.FOR_GOOD]: \\"For Good\\"
};

export default locale;
",
  ],
  Array [
    "dist/src/i18n/locales/zh.ts",
    "import { Locale, K } from \\"../constants\\";

const locale: Locale = {
  [K.FOR_GOOD]: \\"For Good\\"
};

export default locale;
",
  ],
  Array [
    "dist/src/index.spec.ts",
    "import i18next from \\"i18next\\";
import * as kit from \\"@next-core/brick-kit\\";

const spyOnAddResourceBundle = (i18next.addResourceBundle = jest.fn());

jest.spyOn(window.customElements, \\"define\\");

jest.spyOn(kit, \\"getRuntime\\").mockReturnValue({
  registerCustomTemplate: jest.fn(),
  registerCustomProcessor: jest.fn(),
} as any);

jest.mock(\\"./lazy-bricks\\", () => void 0);

// Use \`require\` instead of \`import\` to avoid hoisting.
require(\\"./index\\");

describe(\\"index\\", () => {
  it(\\"should add i18n resource bundle\\", () => {
    expect(spyOnAddResourceBundle).toBeCalled();
  });
});
",
  ],
  Array [
    "dist/src/index.ts",
    "import \\"./i18n\\";
",
  ],
  Array [
    "dist/tsconfig.json",
    "{
  \\"extends\\": \\"../../tsconfig.json\\",
  \\"include\\": [
    \\"./src\\",
    \\"../../declarations\\"
  ]
}
",
  ],
  Array [
    "dist/webpack.config.js",
    "const { bricks, merge } = require(\\"@next-core/webpack-config-factory\\");

const { webpackCommonFactory, webpackDevFactory, webpackProdFactory } = bricks;

module.exports = merge(
  webpackCommonFactory(),
  process.env.NODE_ENV === \\"development\\"
    ? webpackDevFactory()
    : webpackProdFactory()
);
",
  ],
]
`;

exports[`loadTemplate should create a new package of dll 1`] = `
Array [
  Array [
    "dist/webpack.dll.js",
    "const { dll } = require(\\"@next-core/webpack-config-factory\\");

module.exports = dll.webpackDllFactory();
",
  ],
]
`;

exports[`loadTemplate should create a new package of libs 1`] = `
Array [
  Array [
    "dist/jest.config.js",
    "const { jestConfigFactory } = require(\\"@next-core/jest-config-factory\\");

module.exports = jestConfigFactory({
  standalone: true,
  cwd: __dirname,
});
",
  ],
  Array [
    "dist/scripts/post-build.js",
    "const { postBuild } = require(\\"@next-core/build-config-factory\\");

postBuild(\\"libs\\");
",
  ],
  Array [
    "dist/src/index.ts",
    "",
  ],
  Array [
    "dist/tsconfig.json",
    "{
  \\"extends\\": \\"../../tsconfig.json\\",
  \\"include\\": [
    \\"./src\\",
    \\"../../declarations\\"
  ]
}
",
  ],
]
`;

exports[`loadTemplate should create a new package of providers 1`] = `
Array [
  Array [
    "dist/jest.config.js",
    "const { jestConfigFactory } = require(\\"@next-core/jest-config-factory\\");

module.exports = jestConfigFactory({
  standalone: true,
  cwd: __dirname,
});
",
  ],
  Array [
    "dist/rollup.snippets.js",
    "import { rollupFactoryForSnippets } from \\"@next-core/rollup-config-factory\\";

export default rollupFactoryForSnippets();
",
  ],
  Array [
    "dist/scripts/generate-version-file.js",
    "const { createVersionFile } = require(\\"@next-core/build-config-factory\\");

createVersionFile();
",
  ],
  Array [
    "dist/scripts/post-build.js",
    "const { postBuild } = require(\\"@next-core/build-config-factory\\");

postBuild(\\"bricks\\");
",
  ],
  Array [
    "dist/scripts/pre-build.js",
    "const { preBuild } = require(\\"@next-core/build-config-factory\\");

preBuild(\\"providers\\");
",
  ],
  Array [
    "dist/scripts/rename-tar-package.js",
    "const { renameTarPackage } = require(\\"@next-core/build-config-factory\\");

renameTarPackage();
",
  ],
  Array [
    "dist/tsconfig.json",
    "{
  \\"extends\\": \\"../../tsconfig.json\\",
  \\"include\\": [
    \\"./src\\",
    \\"../../declarations\\"
  ]
}
",
  ],
  Array [
    "dist/webpack.config.js",
    "const { bricks, merge } = require(\\"@next-core/webpack-config-factory\\");

const { webpackCommonFactory, webpackDevFactory, webpackProdFactory } = bricks;

module.exports = merge(
  webpackCommonFactory(),
  process.env.NODE_ENV === \\"development\\"
    ? webpackDevFactory()
    : webpackProdFactory()
);
",
  ],
  Array [
    "dist/.gitignore",
    "/src/
",
  ],
]
`;
